balancer-js/src/modules/pools/index.ts

```diff
@@ -27,13 +33,295 @@ import { Simulation, SimulationType } from '../simulation/simulation.module';
 import { PoolGraph } from '../graph/graph';
 import { PoolFactory__factory } from './pool-factory__factory';
 import * as Queries from './queries';
+import {
+  queryJoinParams,
+  JoinExactInParams,
+  JoinExactOutParams,
+  ExitExactOutParams,
+  ExitProportionallyParams,
+  ExitToSingleTokenParams,
+  queryExitParams,
+} from './queries/types';
+import { Stable } from './pool-types/stable.module';
+import { ComposableStable } from './pool-types/composableStable.module';
+import { Weighted } from './pool-types/weighted.module';
+import { MetaStable } from './pool-types/metaStable.module';
+import { StablePhantom } from './pool-types/stablePhantom.module';
+import { Linear } from './pool-types/linear.module';
 import { EmissionsService } from './emissions';
 import { proportionalAmounts } from './proportional-amounts';
 
 const notImplemented = (poolType: string, name: string) => () => {
   throw `${name} for poolType ${poolType} not implemented`;
 };
 
+class WrappedPool implements PoolWithMethods {
```

Comment:
I don't fully understand the need for this extra class. Do you mind giving more details on it?
And I don't have a clear idea on how to implement this, but I was under the impression that we could achieve refreshed pool data by focusing on having the poolsOnChain dataSource updated to a given blockNumber.

Reply:
Got it 👍 
I think what I had in mind is pretty similar with what you have done, but instead of having a new WrappedPool class with its own state, you'd update the state of Pools entirely and then get a new PoolWithMethods from it.
You wouldn't be able to keep the same object with a new state, but would get a new object with a new state. Not sure if that's a big deal though.
And thinking about differences on these approaches, one benefit we would have is to avoid having to duplicate all pool attributes on this new class.

Reply:
So in that case the UI (or user) would have to do something like:
```
let pool = Pools.find(id);
let tx = pool.join(....)
// wait some time
await Pools.refresh()
pool = pools.find(id)
tx = pool.join(...)
```
which I thought felt a bit less user friendly than:
```
const pool = Pools.find(id, blockNo);
const tx = pool.join(....)
// wait some time
pool = pools.find(id, blockNo)
```
Bearing in mind it would be the same for all the other pool methods.
I find it easier to think of the `WrappedPool` as a self contained pool class too - but that could just be a personal thing.

Reply:
Yeah, it does feel more user friendly to be able to pass blockNumber as a parameter on the join method 👍

Reply:
Hey, thank you for the PR. Instead of introducing additional state management to the SDK, I suggest focusing on separating the data fetching logic from the joins, exits, and other services. For instance, the dependency of vaultModel > PoolsSource on a specific data class makes it challenging to replace or mock. It would be better to make it dependent on an interface and then pass an implementation of that interface during instantiation. This approach enables easy replacement with other data sources, which can be independently implemented and reused. Currently, we have two pools repositories with on-chain refreshing.

I'll look into Wrapped class and other things tomorrow to get a better picture.

Reply:
Sounds good. Agreed thats definitely the best solution, just looking at possible quick fixes if that takes a long time to implement. Lets see what your conclusions are.